### ISSUE TITLE
Future interfaces to vectors from other libraries

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#745]
[This is related to #684 but addresses a bigger issue.]

Timo, Bruno and I have brainstormed a bit about how we think the vector classes (and, ultimately, maybe also the matrix classes, but this is for a future discussion) should look like. The fundamental issues are:

1/ PETSc does not provide efficient element-wise access to vectors. It is also not thread-safe. The same is going to be true for the Trilinos Tpetra objects.
2/ It is difficult to ensure synchronisation of elements if they are stored on accelerators, and this limitation is one of the reasons why Tpetra does not allow efficient element-wise access.
3/ We currently have a problem in that we have a single PETSc vector class that is used for both read-write completely distributed vectors, and read-only ghosted vectors. The same is true for the Trilinos wrappers. In other words, the semantics we attach to different kinds of vectors is not carried into the type system.

Here is an idea of how we think this whole mess might be addressed:
1/ Create two sets of parallel vector classes: Ghosted vectors and completely distributed vectors.
2/ The existing member functions of parallel vector classes should neatly fall into two categories: ones that access individual elements in some ways, and ones that treat the vector as a whole (e.g., computing norms, adding vectors, etc). In other words, the two kinds of vectors do not need to duplicate code as they have different and almost completely mutually exclusive member functions.
3/ Ghosted vectors allow reading and writing individual elements, but not exchange of elements across processors. In other words, while they know which part of a global vector they store, they do not need an `MPI_Comm` nor do they know where the other parts are. Ghosted vectors can store overlapped ranges of elements (which would then be ghost elements).
4/ Since they do not allow exchange of data, and since their interface is so small, ghosted vectors don't need to build on PETSc or Trilinos. They should in fact be rather simple to either implement from scratch, or just build on (a stripped-down version of) Martin's `parallel::distributed::Vector` class.
5/ What this means is that we really only need a single `GhostedVector` class. It can be assigned from PETSc or Trilinos vectors, of course, which may require importing elements from other processors.
6/ In almost all of the cases in the library where we currently instantiate stuff for PETSc and Trilinos vectors (e.g., in `DataOut, KellyErrorEstimator, FEValues::get_function_*`) what we really want is a ghosted vector. Using this would drastically reduce the number of instantiations.
7/ Non-ghosted vector classes always act as a whole (from a deal.II perspective). They can be assigned to from a GhostedVector, for example at the end of assembly (which would happen into a GhostedVector). They would indeed need to be different for PETSc and Trilinos (Epetra/Tpetra) as well as classes that we want to build around GPU-support libraries. However, since non-ghosted vectors do not have access to individual elements, the major hurdle to writing such interfaces is gone.
8/ Because we do not allow access to elements of non-ghosted elements, the only times data needs to be synchronized between wherever it is actually stored (locally, remote, or on accelerators) and where users may access it, is in operator=. This greatly simplifies getting the semantic model of which data is current. (Witness our many years of agony trying to define this, until we finally decided on a similar model: namely that ghosted vectors are read-only.) It also allows libraries to do what's efficient (e.g., not store data on host memory and only on the GPU), and it makes it easy for us to make access to elements thread-safe (which is very difficult or impossible right now).

For example, we may think of interfaces such as this (that's basically all we need):

```
  class GhostedVector {
    void reinit (const IndexSet &locally_owned_elements,
                     const IndexSet &locally_relevant_elements);   // could in principle do with just one 
                                                  // index set (like Epetra) but may be more convenient for later
                                                  // transfer to non-ghosted vectors to store both

    double operator() (uint...);  // read-write
    void set (...);    // set multiple elements at once
    void add (...);    // add multiple elements at once
    void get (...);    // get multiple elements at once

   operator= (PETScWrappers::Vector &);
   operator= (TrilinosWrappers::Vector &);
  };

namespace TrilinosWrappers {
  class DistributedVector {
     void reinit (const IndexSet &locally_owned_elements);   // note: only one index set
     operator + ();
     operator += ();
     double l2_norm();
     // etc
  };
}
```

**Comment by kronbichler** [comment]
Here a few thoughts from my point of view:
- I like the idea of a clear separation between the deal.II infrastructure and external vector interfaces. In particular with respect to things like KellyErrorEstimator, FEValues::get_function_* and so on. So :+1: for that. As suggested, this would share some code with parallel::distributed::Vector but in order to fulfill the requirements above, several things would need to be stripped out, such as

**Comment by Rombur** [comment]
* I like the idea of using the ghosted vector in serial. It would make it so simple to change a program to use MPI
* I don't think that this interface will lock us in using CPU. We could use a library like ViennaCL to represent the underlying data of the ghosted vectors. Thus we will only have to change the ghosted vector but we should be able to keep interface. We will still have a problem when we have to transfer data from one library to the other (let's say from ViennaCL to Tpetra) but I don't see a way around it.

**Comment by tjhei** [comment]
> how to organize the indices in ghosted vectors

I was arguing along those lines (represent everything in deal.II using local indices) when chatting with Bruno and Wolfgang but this would require some deeper changes inside deal.II and they were not convinced it is worth it (only small performance win; amount of work ...). I think we should have this discussion though.

The main change would be to have DoFHandler return local indices in ```get_dof_indices``` (maybe when constructed with a certain flag to be backwards compatible). This should be enough to make things like DataOut work without any internal changes. Of course we need to take a closer look at ```ContraintMatrix``` etc.. To me this doesn't look like it would be very hard to do.

One argument for keeping things as is: it is very intuitive to think in global indices because you construct a single large system like in serial code.

> Could we go one step further and make the ghosted vector the only way to enter the above deal.II functions, i.e., also remove dealii::Vector instantiations?

Are you suggesting that the ```dealii::Vector``` should also require creating a ```GhostedVector```? I like that serial/parallel code becomes identical but would hate for step-1 to 6 to become more complicated.

**Comment by kronbichler** [comment]
> Are you suggesting that the dealii::Vector should also require creating a GhostedVector? I like that serial/parallel code becomes identical but would hate for step-1 to 6 to become more complicated.

Thinking about this some more, I see two variants which would provide some benefit:
- Take the path described by @bangerth in the original post, i.e., introduce GhostedVector alongside with our distributed vectors which are reduced to a minimal configuration, though.
- Taking @tjhei and my suggestion of keeping local dof indices in most data structures allows for a very neat code structure: The ghosted vectors would simply be dealii::Vector and we always access them through local indices. The advantage is that serial and parallel codes are very similar except for the solvers and matrices where we need to have communication and 1-to-1 distribution of indices. No problems introduced for all non-MPI programs. The disadvantage is the ambiguity of indices from different processors' views and, possibly, the desire to have different 'local' enumerations, e.g. by just having the locally active dofs vs. the locally relevant dofs for example.

**Comment by tjhei** [comment]
I would think the local enumeration would need to be owned, active, then relevant enumerated from 0...n_owned+n_active+n_relevant. Otherwise operations like DataOut won't work.

**Comment by bangerth** [comment]
On @kronbichler 's points:

2/ "Can we go one step further and make the ghosted vector the only way to enter the above deal.II functions, i.e., also remove dealii::Vector instantiations": I thought about that as well, but couldn't get warm with the idea. The basic fact is that most people don't use deal.II in parallel and what concerns me is that we would have to spend a significant amount of time in the early tutorials explaining why exactly we go to the length to introduce two different vector classes. If you're coming from the Matlab world, this is going to present a significant hurdle if all you want to do is understand step-3. To me, clarity and simplicity in education are an important point.

5/ "Another point I want to discuss is how to organize the indices in ghosted vectors." (a.k.a local vs global numbering): @tjhei argued for local numbering as well. My view is that this would require a massive undertaking because it would be all pervasive. But my main objection is again the education part: having a single, global linear system indexed with global indices has a very nice intellectual clarity to it. It is easy to understand; it is so much easier to debug problems because you can look at vector elements from different processors and know how they relate; etc. On this issue, I think that if the current way has inefficiencies, then we should look into the data structures we use. For example, if the ghosted vectors we will use are organized in a way where the largest contiguous block of unknowns (typically the locally owned ones) comes first in storage, then we can get random access at O(1) effort for the large majority of accesses (with two comparisons of an index *i* against the lower and upper bound of the first block of unknowns, following by a single subtraction of the lower bound). I believe that the `parallel::distributed::Vector` already uses such a scheme. (Petsc also does, for that matter.) I would be very surprised if that was not enough.

So I have to admit that I really like the clarity of global numbering and am willing to work on other parts of the code to retain it, rather than write massive amounts of difficult documentation.

**Comment by Rombur** [comment]
On @kronbichler 's point number 2: I don't think that it would make the serial tutorial more complicated. I don't think that's harder to understand than having to build a SparsityPattern to use a SparseMatrix.  We could also delay the detailed explanation of the ghosted vector for the MPI tutorials. I think that simplifying the switch from a serial code to a parallel code is worth the small hassle of having to write in a ghosted vector in serial. In my (limited) experience, people write their code in serial, use it for a few test cases, and then, they realize that they can't do a more realistic case because they need to rewrite the code for MPI. Making it easier to switch to MPI would help people down the line.

**Comment by kronbichler** [comment]
I do agree on the educational aspect of using the simple dealii::Vector for serial programs as suggested by @bangerth.

*Regarding local vs. global indices:* The point of my last post was that using local indices throughout with dealii::Vector only does indeed have educational value too because this approach allows you to easily switch between serial and MPI computations. All you change is the layout of the matrix (you equip it with means to transform local into global indices or whatever our external interface want) and by copying + combining accumulated data into a distributed vector with communicator. However, I also agree on the fact that this requires a lot of work because we need to change the matrix interface, vector interfaces. At least as big a change as the switch to 64-bit indices. So I guess @bangerth is right that this boat has sailed.

On the other hand, I would see the value of storing local indices *inside* the library: Whenever we have cases where efficiency matters, we could resort to GhostedVector::local_element approach. The place where this is done already is the matrix-free framework where everything else than processor-local indices for vector access kills performance. Heck, the DG implementation that we will hopefully merge in the next few months already spends 60% of computational time on vector access with direct array access and I am experimenting with vectorized reads because of that bottleneck. My plan in the longer run is to get the dof management via suitable modifications in DoFHandler instead of what is now internal::MatrixFreeFunctions::DoFInfo. Having two codes for the same purpose doubles maintenance burden. My plan for that would necessitate DoFHandler to hold indices in local space, possibly even compress contiguous ones, and having dealii::Vector and GhostedVector to access data with local_element(index) or, cleaner, vector.begin()+index.

That model does not force changes to the externally visible interface, because optimized read operations from vectors would be internal to the library. While initially used in a limited context, we could provide fast access to vector entries wherever we need that, e.g. in `FEValues::get_function_xxx`. We could also cheaply return global dof indices with `DoFAccessor::get_dof_indices`. Using 'compare + add' in the largest chunk and the usual IndexSet stuff otherwise, i.e., the inverse of what happens for vector access in global indices and what is fine as long as you don't do _linear algebra_ with that data. The additional benefit is that 32-bit integers should be enough for the local indices for a long time to come, cutting the memory of DoFHandler into almost half for 64 bit integers.

But back to topic: If we feel that keeping globalized indices in user interfaces is the right thing to do (and it probably is, given the huge amount of resources necessary to change that throughout), I think that we can easily work on the necessary infrastructure in the background where bottlenecks come up. Having a unified interface with GhostedVector + deal.II vector for everything that belongs to deal.II really is working towards that goal in my opinion. I have not really thought about what would happen if we need want external libraries to do data management of the local computations as I discussed with @Rombur (say using GPUs for assembly or other fancy stuff inside deal.II via Trilinos Kokkos). But that is likely a relatively orthogonal requirement to the MPI case, and not too difficult to handle when we have a single processor-local model via GhostedVector.

**Comment by bangerth** [comment]
I've looked through this thread again and it seems that we all agree that the 
original proposal is probably going in the right direction. In particular, 
this would include:

1/ Creating a mental model in which there are two kinds of vectors
- a kind (called GhostedVector) that (i) stores a subset of all elements,
   (ii) allows element access, (iii) knows nothing about similar objects on
   other processors, (iv) does not allow global vector operations such as
   taking the norm.
- a kind (called DistributedVector) that (i) knows about a parallel universe,
   (ii) stores disjoint subsets of elements, (iii) can do global operations
   in parallel, (iv) does not allow access to individual elements.

2/ Our interfaces to Petsc, Trilinos, ... would only provide wrappers for the 
DistributedVector model. These should actually be relatively simple to write 
as all we would need to do is remove existing functionality.

3/ We would only provide two classes that model the GhostedVector concept. 
This would be:
- A stripped down version of parallel::distributed::Vector (which would lose
   all MPI functionality and the distinction between locally owned and
   ghost elements but retains the ability to store a subset of elements).
- The current Vector class (which, incidentally, models both kinds of
   concepts for the special case that the universe exists of a single
   processor and consequently, all elements are stored locally).
One could imagine actually using only one class for these. However:
- I like the clarity of our two-model concept
- If we wanted to rewrite the current Vector class, we could simply
   derive it from the stripped down p::d::Vector class and provide
   it with the vector-vector operations.


Martin and Bruno make good points about other issues, such as letting the 
DoFHandler store local (in addition to global) DoF indices. I think this is a 
worthwhile addition and should be easy to achieve (let the DoFLevels objects 
store *local* dof indices, and provide a local-to-global translation table 
somewhere that one has to use in all of those places where we currently read 
from the DoFLevels). But, these things are orthogonal -- let us discuss them 
in a separate issue.

**Comment by bangerth** [comment]
I think the design outlined above was considered useful in today's discussion at the workshop if maybe not met with universal approval (@tjhei had good points, even though I may have disagreed).

Here's the mockup we were talking about in the group discussion:
```
namespace LinearAlgebra
{
  class ReadWriteVector           // to be based on parallel::distributed::Vector
  {
    public:
      ReadWriteVector (IndexSet &read_write_set);
      ReadWriteVector (IndexSet &read_write_set,
                                   IndexSet &read_set);               // ?? I forgot why we wanted this?

      double & operator() { ... }
  };


  class VectorSpaceVector
  {
    public:
      virtual operator = (const ReadWriteVector &) = 0;
      virtual operator+= () = 0;
      virtual operator* (const double) = 0;
  };


  namespace Trilinos
  {
    namespace Epetra
    {
      class Vector : VectorSpaceVector
      {
         virtual copy_from (const ReadWriteVector &,
                                     CombineMode) {...};
         virtual operator+= () {...};
         virtual operator* (const double) {...};
      };


      class SparseMatrix
      {
        void vmult (Vector &dst,
                          const Vector &src);
      };
    }
  }

}

// to replace the existing dealii::Vector class. will contain the remaining
// member functions currently in parallel::distributed::Vector
class Vector :
  LinearAlgebra::ReadWriteVector,
  LinearAlgebra::VectorSpaceVector
{
    virtual operator+= () {...};
    virtual operator* (const double) {...};
};


void solve (ReadWriteVector &solution, ReadWriteVector rhs)
{
  Trilinos::Vector b = rhs, x = solution;
  Trilinos::solver (A, x, b);
  solution = x;
}
```

**Comment by Rombur** [comment]
Following the discussion in #1514, we should add a new VectorView class that does not own the memory. This would allow almost instant wrapping among different libraries without costly copies back and forth.

**Comment by tjhei** [comment]
I think it would be better if ReadWriteVector would support optionally owning the memory. Otherwise we have two different vector types that we need to have instantiations for.

**Comment by bangerth** [comment]
You can do this internally via something like this:
```
  class ReadWriteVector
  {
    private:
     struct DataManager {
       virtual ~DataManager () = 0;
       virtual double * get_data_pointer () = 0;
       virtual void destroy_data ();
    };

    std::unique_ptr<DataManager> data_manager;

    struct LocallyOwnedData : DataManager {
      ... implemented to allocate the data;
      virtual void destroy_data () { delete data; }
   private:
      double *data;
   }

    struct TrilinosVectorView : DataManager {
      ... implemented to not own data, but create a view into a Trilinos vector ....;
      virtual void destroy_data () { /* do nothing, Trilinos continues to own data */ }
   }

    double *values;


   public:

    ReadWriteVector (...)
      :
      data_manager (new LocallyOwnedData(...)),
      values (data_manager.get_data_pointer())
    {}

    ReadWriteVector (TrilinosWrappers::MPI::Vector)
      :
      data_manager (new TrilinosVectorView(...)),
      values (data_manager.get_data_pointer())
    {}

  }
```

This way (i) you need only one class, but it dispatches its memory management operations internally (which is slower than via templates, but it doesn't matter because we don't do this often), and (ii) all data accesses happen via a pointer to memory we know where it is, so it's fast.

**Comment by Rombur** [comment]
Now that the new interface for Trilinos vector has been merged, only PETSc vector and our own distributed vector still need to implement the new interface. I have never used PETSc so I don't want to implement it. For deal.II vector, I don't know if you have time to do it @kronbichler. Otherwise, I can do it. Once this is done, I will go back to the classes that I have already implemented and try to optimize the code. I haven't done it yet since the interface was not stable.

**Comment by kronbichler** [comment]
I will try to do this next week.

**Comment by Rombur** [comment]
@kronbichler Do you need help on this?

@davydden @tjhei Can one of you implement the PETSc interface? I don't really want to spend time learning PETSc since I never used it.

**Comment by kronbichler** [comment]
@Rombur sorry for the delay, I will start a simple version that simply puts the interface in place and avoids using VectorView next week. The rest of the discussion in #2535 (lightweight conversion to/from ReadWriteVector) takes a bit more time (and has prevented me from starting this) will follow later.

**Comment by Rombur** [comment]
What is the next step for the integration of the new vector interface? #3073 shows that there is still a lot of work to do before we can fully use the new vectors. Should we just try to use the new vector with some random classes until everything works or is there a better way to do that?

**Comment by bangerth** [comment]
Could we just add the new vector classes to the list of vectors we instantiate templates with? This would give the whole thing a fairly good compile-time shake-out at least.

**Comment by Rombur** [comment]
There are several places where we use ```operator()``` Is the preferred method to make the class a friend or should we use a ```ReadWriteVector```?

**Comment by bangerth** [comment]
I think the question is whether you are implementing an algorithm that sees the vector as a collection of individual entries (say, `VectorTools::interpolate`) in which case you want to go through `ReadWriteVector::operator()`, or if you see the vector as an element in a vector space (say, in `SparseMatrix::vmult`) in which case you will have to access *all* elements and the preferred way would be to make the function or surrounding class a `friend` of the vector-space vector class.

**Comment by bangerth** [comment]
In other words, my long-term plan would be to convert basically everything outside the `lac/` subdirectory to not take concrete vectors as a template argument, but instead take a reference to a `ReadWriteVector`. The compiler would then cast the argument from whatever implementation of the `ReadWriteVector` in `newlinearalgebra::Vector` (or any other classes, if we chose to write some) to the base class. (Or if we implement the actual data storage scheme in `ReadWriteVector`, which may already be what we do, then there is no need for any cast).

Does that make any kind of sense?

**Comment by kronbichler** [comment]
As noted in #3656 and #4043, we need to be aware of the situation where the MPI communicator underlying a vector has been deleted by `MPI_Comm_free`. Of course, all vectors using this communicator should be deleted at this point, but we cannot ensure this when the vector and MPI interfaces are controlled in third party libraries - and possibly not even without our code in `LinearAlgebra::distributed::Vector` without tediously keeping track of things via `SmartPointer`. While we can request a user to delete all vectors prior to using `MPI_Comm_free` (explicitly or implicitly within deal.II in `parallel::Triangulation`), there might be stale states left in `GrowingVectorMemory`.

In order to avoid problems, we need to make sure that `operator=` and `reinit()` functions of the new vector classes **never** try to retain the old state (and communicator) of the vector.

**Comment by bangerth** [comment]
Here is a short note about a separate topic: Eventually, we want to base the Trilinos interfaces on the newer Tpetra package (https://trilinos.org/packages/tpetra/). I had some conversations about the stability of Tpetra APIs at this point with Mark Hoemmen -- the main Tpetra developer -- and these were his answers:

.....................................
> We're looking forward to using Tpetra for deal.II. Do you think it's at a 
> point where there is largely API stability so we don't have to write and 
> rewrite our code multiple times?

The interface for thread-parallel sparse graph and matrix construction and modification – what I call the “fill interface” – will continue to evolve.  However, other parts of the interface will not change much. 

Let me elaborate on the fill interface.  Tpetra’s current fill interface has only limited support for thread safety and scalability.  For example, CrsMatrix::sumInto*Values is thread safe and scalable (in so far as hardware atomic += thread-scales), but only on the CPU, not on the GPU.  CrsGraph and CrsMatrix methods that change the graph structure are NOT thread safe.

If you want thread-parallel sparse graph and matrix construction and modification, here is the most efficient thing you could do:

1. Split up “local” (per MPI process) and “global” (across MPI processes) assembly
2. Do local work using Kokkos data structures (e.g., KokkosSparse::CrsMatrix, which has sumIntoValues and replaceValues methods)
3. Pass them off to Tpetra data structures and do global work, e.g., using doExport for global assembly

Kokkos has examples, esp. the “FENL” (finite-element assembly and solve) example, that illustrate thread-parallel local graph construction and local assembly.

Our plan is to deprecate and remove the fill interface in Tpetra::CrsGraph and Tpetra::CrsMatrix.  Local things are local; global things are global.  We will continue to support the interface for a while.

I will be happy to work with you and/or your folks in order to make thread-parallel fill better ☺

> On a different note, given that Kokkos has all of these placement and policy 
> parameters, how do you deal with these in Tpetra? Do you run them all the way 
> through the user interface and let Tpetra users choose them? Or do you make 
> educated guesses and hide this choice from the user?

Tpetra asks users to specify the Kokkos execution space.  It uses that execution space’s default memory space as its memory space.  Tpetra objects will eventually all behave like Kokkos::DualView (MultiVector and Vector already do now); this will help gradual porting to use GPUs, and will also let users decide whether to run in HBM or main memory on KNL.  Otherwise, Kokkos developers’ view is that special placement and policy parameters, like texture fetch (RandomRead), belong only at the entry point of computational kernels.  Use mostly undecorated Views wherever possible, and only decorate on entry to a computational kernel, like sparse matrix-vector multiply.  The point is to avoid building kernels and code for too many different template parameter combinations.  We have taken this approach in the design of KokkosKernels.  I would be happy to walk you through examples if you like.
........................

**Comment by Rombur** [comment]
Since #3957 has been merged, I think the new interface is ready to be used. The petsc wrappers are still missing but it should be easy for someone who knows petsc to add them by looking at the trilinos wrappers.

**Comment by drwells** [comment]
@Rombur would it be possible for you to post the slides you showed yesterday?

**Comment by Rombur** [comment]
I cannot post the slides (I would need to get the authorization from ORNL) but I will post a summary (yes, it does make a difference...)

**Comment by drwells** [comment]
Fair enough (that is the way these things go with the labs).

**Comment by kronbichler** [comment]
I think we can consider this fixed by #15197. We can re-open again if we think we have new ideas, but I like the current state and believe that local improvements can be the way forward.
